### PR TITLE
Add D* Lite.

### PR DESCRIPTION
**Pull request by vss2sn**
- [Done] This project only accept codes for python 3.9 or higher.
- [Done] If you add a new algorithm sample code, please add a unit test file under `test` dir.
     This sample test code might help you : https://github.com/AtsushiSakai/PythonRobotics/blob/master/tests/test_a_star.py
- [N/A] If you fix a bug of existing code please add a test function in the test code to show the issue was solved.
- [Done] Please fix all issues on CI (All CI should be green), before code review.

Add D* Lite algorithm to the repository. It contains the live version of the algorithm, which detects obstacles as it moves, and re-routes. The creating/detection of the obstacles can be random or specified by stating at which time step the obstacle is to be detected. The function names used and the structure of the code attempt to mirror the pseudocode from the papers linked.

#### Additional information

Awaiting feedback.

### CODE REVIEW

File: PathPlanning/DStarLite/DStarLite.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
+"""
+D* Lite grid planning
+author: vss2sn (28676655+vss2sn@users.noreply.github.com)
+Link to papers:
+D* Lite (Link: http://idm-lab.org/bib/abstracts/papers/aaai02b.pd)
+Improved Fast Replanning for Robot Navigation in Unknown Terrain
+(Link: http://www.cs.cmu.edu/~maxim/files/dlite_icra02.pdf)
+Implemented maintaining similarity with the pseudocode for understanding
+"""
+from __future__ import annotations
```
Comment (AtsushiSakai) [inline PathPlanning/DStarLite/DStarLite.py:10]: This project only support Python 3.9. So, maybe this import is not needed.

File: PathPlanning/DStarLite/DStarLite.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+
+    motions = [
+        Node(1, 0, 1),
+        Node(0, 1, 1),
+        Node(-1, 0, 1),
+        Node(0, -1, 1)
+    ]
+
+    def __init__(self, x_max: int, y_max: int, obstacles: list,
+                 spoofed_obstacles: list = list()):
```
Comment (AtsushiSakai) [inline PathPlanning/DStarLite/DStarLite.py:49]: list should not used for an argument: Ref: https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

File: PathPlanning/DStarLite/DStarLite.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    def h(self, s: Node):
+        # Modify if allowing diagonal motion
+        return abs(self.start.x - s.x) + abs(self.start.y - s.y)
+
+    def calculate_key(self, s: Node):
+        return (min(self.g[s.x][s.y], self.rhs[s.x][s.y]) + self.h(s)
+                + self.km, min(self.g[s.x][s.y], self.rhs[s.x][s.y]))
+
+    def is_valid(self, node: Node):
+        if node.x >= 0 and node.y >= 0 and node.x < self.x_max and \
```
Comment (AtsushiSakai) [inline PathPlanning/DStarLite/DStarLite.py:100]: `if 0 <= node.x < self.x_max and 0 <= node.y < self.x_max:`
is better?

File: PathPlanning/DStarLite/DStarLite.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    obsy = [1, 2]
+    obstacles = [Node(x, y) for x, y in zip(obsx, obsy)]
+
+    # Obstacles discovered at time = row
+    # time = 1, obstacles discovered at (0, 2), (9, 2), (4, 0)
+    # time = 2, obstacles discovered at (0, 1), (7, 6)
+    # ...
+    spoofed_obstacles = list()
+    spoofed_obstacles_x = [[0, 9, 4], [0, 7], [], [], [], [], [], [5]]
+    spoofed_obstacles_y = [[2, 2, 0], [1, 6], [], [], [], [], [], [4]]
```
Comment (AtsushiSakai) [inline PathPlanning/DStarLite/DStarLite.py:254]: Could you use the same configuration of astar.py? 
https://github.com/AtsushiSakai/PythonRobotics/blob/1de0ffddb1ea11396a019403bbb986c8df8e0bdd/PathPlanning/AStar/a_star.py#L236-L263

**Review by AtsushiSakai — CHANGES_REQUESTED** [review]
Thank you for great PR!! I have some small comments.

File: PathPlanning/DStarLite/DStarLite.py (hunk 1 above)
Comment (vss2sn) [reply PathPlanning/DStarLite/DStarLite.py:10]: That was left as Python 3.8 is currently the default for Ubuntu, but will modify.

File: PathPlanning/DStarLite/DStarLite.py (hunk 2 above)
Comment (vss2sn) [reply PathPlanning/DStarLite/DStarLite.py:49]: Will modify.

File: PathPlanning/DStarLite/DStarLite.py (hunk 3 above)
Comment (vss2sn) [reply PathPlanning/DStarLite/DStarLite.py:100]: Yes, that works.

File: PathPlanning/DStarLite/DStarLite.py (hunk 4 above)
Comment (vss2sn) [reply PathPlanning/DStarLite/DStarLite.py:254]: I can modify the start and end coordinates.

Some clarifications/context:
- **Negative grid coordinates**: The grid currently only takes indices >= 0 and makes a 1 to 1 correlation between grid coordinate and index for easy understanding.
- **Robot radius**: Adding robot radius will make it a little more complex to check for obstacles, and attempting simplicity
- **Boundary wall**: The boundary wall is not needed as the planner will not attempt to expand points not in the grid (though I can add in a check to ensure start and end points are on the grid)
- **Obstacles**: The obstacles were chosen specifically for forcing a reroute during the dynamic planning, and demonstrate how an example of discovering obstacles with time

Please let me know which of the changes referred to above (in bold) you would like me to make; will make them.

**Comment by vss2sn** [comment]
Attached gif for README.md. 
DStarLite

**Comment by AtsushiSakai** [comment]
Thank you for updating the code.
The code looks good to merge, but  I have a question about the animation.
You code rerouted the path after new obstacle are detected.
Is it possible to show the `before rerouting`, not only `after rerouting`?
I think that makes more sense that D* Lite has rerouting feature.

**Comment by vss2sn** [comment]
Modified the code to display the paths; displays the path before the last reroute and the current valid path. Not displaying all previous paths as with more than a single reroute, it will get difficult to discern the different paths.

DStarLite

**Review by AtsushiSakai — APPROVED** [review]
LGTM. Thank you for great PR!! I will merge it. @vss2sn

**Pull request merged**